pulsar-broker/src/main/java/com/yahoo/pulsar/broker/loadbalance/impl/DeviationShedder.java

```diff
@@ -0,0 +1,134 @@
+package com.yahoo.pulsar.broker.loadbalance.impl;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeSet;
+
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+import com.yahoo.pulsar.broker.BrokerData;
+import com.yahoo.pulsar.broker.ServiceConfiguration;
+import com.yahoo.pulsar.broker.loadbalance.LoadData;
+import com.yahoo.pulsar.broker.loadbalance.LoadSheddingStrategy;
+
+/**
+ * An abstract class which makes a LoadSheddingStrategy which makes decisions based on standard deviation easier to
+ * implement.
+ */
+public abstract class DeviationShedder implements LoadSheddingStrategy {
+    // A Set of pairs is used in favor of a Multimap for simplicity.
+    protected TreeSet<Pair<Double, String>> metricTreeSetCache;
+    protected TreeSet<Pair<Double, String>> bundleTreeSetCache;
+
+    /**
+     * Initialize this DeviationShedder.
+     */
+    public DeviationShedder() {
+        bundleTreeSetCache = new TreeSet<>();
+        metricTreeSetCache = new TreeSet<>();
+    }
+
+    // Measure the load incurred by a bundle.
+    protected abstract double bundleValue(String bundle, BrokerData brokerData, ServiceConfiguration conf);
+
+    // Measure the load suffered by a broker.
+    protected abstract double brokerValue(BrokerData brokerData, ServiceConfiguration conf);
+
+    // Get the threshold above which the standard deviation of a broker is large
+    // enough to warrant unloading bundles.
+    protected abstract double getDeviationThreshold(ServiceConfiguration conf);
+
+    /**
+     * Recommend that all of the returned bundles be unloaded based on observing excessive standard deviations according
+     * to some metric.
+     * 
+     * @param loadData
+     *            The load data to used to make the unloading decision.
+     * @param conf
+     *            The service configuration.
+     * @return A map from all selected bundles to the brokers on which they reside.
+     */
+    @Override
+    public Map<String, String> selectBundlesForUnloading(final LoadData loadData, final ServiceConfiguration conf) {
```

Comment:
not sure, do we think `findBundlesForUnloading` will be better as we don't select from given input?